src/Simplex/Messaging/Agent/Client.hs

```diff
@@ -529,41 +532,49 @@ getSMPServerClient c@AgentClient {active, smpClients, msgQ} tSess@(userId, srv,
           unless (null conns) $ notifySub "" $ DOWN srv conns
           unless (null qs) $ do
             atomically $ mapM_ (releaseGetLock c) qs
-            unliftIO u $ reconnectServer c tSess
+            unliftIO u $ resubscribeSMPSession c tSess
 
         notifySub :: forall e. AEntityI e => ConnId -> ACommand 'Agent e -> IO ()
         notifySub connId cmd = atomically $ writeTBQueue (subQ c) ("", connId, APC (sAEntity @e) cmd)
 
-reconnectServer :: AgentMonad m => AgentClient -> SMPTransportSession -> m ()
-reconnectServer c tSess = newAsyncAction tryReconnectSMPClient $ reconnections c
+resubscribeSMPSession :: AgentMonad m => AgentClient -> SMPTransportSession -> m ()
+resubscribeSMPSession c@AgentClient {smpSubWorkers} tSess =
+  atomically (getTSessVar tSess smpSubWorkers) >>= either newSubWorker (\_ -> pure ())
   where
-    tryReconnectSMPClient aId = do
+    newSubWorker v = do
+      subWorkerId <- atomically $ stateTVar (workerSeq c) $ \next -> (next, next + 1)
+      subWorkerAsync <- async $ runSubWorker subWorkerId `E.catchAny` const (atomically $ cleanup subWorkerId)
+      atomically $ putTMVar v SubWorker {subWorkerId, subWorkerAsync}
```

Comment:
is there something that prevents multiple workers for the same session with different IDs? Check for session and adding to map are in separate atomically. Or all but one are killed afterwards? I may be missing something..

Reply:
killed in `cleanup`?

Reply:
It is `getTSessVar` that does the reservation by putting empty TMVar at the key. `putTMVar` here only fills it it with the actual value.

Reply:
right, thanks 🤦‍♂️